### PR TITLE
nb: Don't display delete button if only one cell.

### LINKED ISSUE #323: Notebook allows last cell to delete
When inside a notebook, the last cell gives a user the option to delete the cell.
screen shot 2018-03-05 at 3 44 19 pm
When the user clicks delete, the notebook goes into an error state that cannot be recovered without refreshing:
screen shot 2018-03-05 at 3 44 26 pm
Expected experience is to not have the delete option on the last cell in a notebook.

### PR DESCRIPTION
**Pull request by ry**
Fixes #323.

Also improve MockDB and title editing.

(Part 1 of 2 in the nb cell refactor)

### CODE REVIEW

File: website/db.ts [hunk 1]
```diff
@@ -225,30 +231,33 @@ export class DatabaseMock implements Database {
...
+    for (const cb of this.authChangeCallbacks) {
+      cb(this.currentUser);
+    }
   }
 
   subscribeAuthChange(cb: (user: UserInfo) => void): UnsubscribeCb {
     this.inc("subscribeAuthChange");
-    return null;
+    this.authChangeCallbacks.push(cb);
+    return () => {
```
Comment (piscisaureus) [inline website/db.ts]: What's this return value used for? It doesn't seem to do what the name of the return type suggests (rather than unsubscribe one listener, it clears out all the listeners)

File: website/nb_test.ts [hunk 2]
```diff
@@ -23,128 +17,147 @@ testBrowser(function notebook_NotebookRoot() {
...
+  const deleteButton = document.querySelector(".delete-button");
+  assert(deleteButton == null);
+});
 
-  title = document.querySelectorAll("div.title > h2");
-  assert(1 === title.length);
-  assert("New Title" === title[0].innerHTML);
-  */
+// Call this to ensure that the DOM has been updated after events.
+function flush() {
```
Comment (piscisaureus) [inline website/nb_test.ts]: Can you give this function a return type?

File: website/nb_test.ts [hunk 3]
```diff
@@ -23,128 +17,147 @@ testBrowser(function notebook_NotebookRoot() {
...
+  rerender();
```
Comment (piscisaureus) [inline website/nb_test.ts]: Is rerender() synchronous?

File: website/nb.ts [hunk 4]
```diff
@@ -637,17 +648,17 @@ export class Notebook extends Component<NotebookProps, NotebookState> {
 
       const titleEdit = h("div", { class: "title" },
         h("input", {
-          class: "title-input",
-          onChange: event => this.onTypedTitle(event),
-          value: doc.title
+          "class": "title-input",
```
Comment (piscisaureus) [inline website/nb.ts]: Why did you add quotes?

File: website/nb.ts [hunk 5]
```diff
@@ -225,10 +225,21 @@ export class Cell extends Component<CellProps, CellState> {
...
+  // component if the props have changed (e.g. if props.onDelete has been
+  // changed as demoed in the notebook_DeleteLastCell test).
   // Code updates are done in componentWillReceiveProps.
-  shouldComponentUpdate() {
-    return false;
+  // TODO This is very hacky. Ideally we wouldn't have to mange CM's
+  // state like this.
+  shouldComponentUpdate(nextProps, nextState) {
+    const propChanged = (name) =>
+      (this.props[name] == null && nextProps[name] != null) ||
```
Comment (piscisaureus) [inline website/nb.ts]: what about `(this.props[name] == null) !== (nextProps[name] == null)`

File: website/db.ts [hunk 6]
```diff
@@ -225,30 +231,33 @@ export class DatabaseMock implements Database {
...
+      this.authChangeCallbacks = [];
+    };
   }
 }
 
-// Default to a mock, so none of the functions errors out during operation in
```
Comment (piscisaureus) [inline website/db.ts]: Happy to see this gone!

File: website/nb.ts (hunk 4 above)
Comment (piscisaureus) [reply website/nb.ts]: Nonetheless class is perfectly valid as an object key. And "value" and "onChange" are definitely not keywords.

Comment (ry) [reply website/nb.ts]: Fixed.

**Comment by piscisaureus** [comment]
Lgtm after squashing
